base/sort.jl

```diff
@@ -533,75 +537,92 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
     return v
 end
 
-# selectpivot!
+# select a pivot for QuickSort
 #
-# Given 3 locations in an array (lo, mi, and hi), sort v[lo], v[mi], v[hi]) and
-# choose the middle value as a pivot
-#
-# Upon return, the pivot is in v[lo], and v[hi] is guaranteed to be
-# greater than the pivot
+# This method is redefined to rand(lo:hi) in Random.jl
+# We can't use rand here because it is not available in Core.Compiler and
+# because rand is defined in the stdlib Random.jl after sorting is used in Base.
+select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash((lo, hi)) % (hi-lo+1)) + lo
```

Comment:
```suggestion
select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash(lo) % (hi-lo+1)) + lo
```
It seems unnecessary to hash tuple of both `(lo,hi)` because of the following mod operation (length `(hi-lo+1)` changes over time). Thus, the pivot selection can be sped up, as demonstrated using v1.8.1
``` julia
julia> using BenchmarkTools

julia> select_pivot(lo::Integer, hi::Integer) = typeof(hi-lo)(hash((lo, hi)) % (hi-lo+1)) + lo
select_pivot (generic function with 1 method)

julia> @btime select_pivot(a,a+b) setup=((a,b) = rand(1:10^6, 2));
  7.441 ns (0 allocations: 0 bytes)

julia> select_pivot_fast(lo::Integer, hi::Integer) = typeof(hi-lo)(hash(lo) % (hi-lo+1)) + lo
select_pivot_fast (generic function with 1 method)

julia> @btime select_pivot_fast(a,a+b) setup=((a,b) = rand(1:10^6, 2));
  3.790 ns (0 allocations: 0 bytes)
```

Reply:
You have a fast computer! I get 13 and 10 ns respectively. But that is still a clear improvement and your reasoning w.r.t. the mod operation is sound. Thanks!

Reply:
That's surprising. I have `CPU: 4 × Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz` MacBook (only two physical cores) which should theoretically be 1.6x faster, not 1.7-2.6x slower. I've shared my system benchmarks [here](https://github.com/IanButterworth/SystemBenchmark.jl/issues/8#issuecomment-1197475145).